daemon/algod/api/server/v2/handlers.go

```diff
@@ -1153,41 +1153,30 @@ func (v2 *Handlers) GetApplicationBoxes(ctx echo.Context, applicationID uint64,
 	castedMax := nilToZero(params.Max)
 	maxBoxThreshold := v2.Node.Config().MaxAPIBoxPerApplication
 
-	// compute `maxReturnNum` from `castedMax` and `maxBoxThreshold := v2.Node.Config().MaxAPIBoxPerApplication`
-	// `maxReturnNum == 0` for returning everything
-	// - `castedMax == 0` and `maxBoxThreshold == 0`,
-	//       return everything
-	// - `castedMax == 0` and `maxBoxThreshold > 0`,
-	//       returns up to `maxBoxThreshold`, check if exceeds
-	// - `castedMax > 0` and `maxBoxThreshold == 0`,
-	//       return up to `castedMax`
-	// - `castedMax > 0` and `maxBoxThreshold > 0`,
-	//       return up to min(castedMax, maxBoxThreshold),
-	//       if return size is `maxBoxThreshold`, check if exceeds
-
-	var boxKeys []string
-	var err error
-
-	dominatedByQryParams := castedMax > 0 && (maxBoxThreshold >= castedMax || maxBoxThreshold == 0)
-	returnsAll := castedMax == 0 && maxBoxThreshold == 0
-
-	if dominatedByQryParams || returnsAll {
-		boxKeys, err = ledger.LookupKeysByPrefix(lastRound, keyPrefix, castedMax)
-		if err != nil {
-			return internalError(ctx, err, errFailedLookingUpLedger, v2.Log)
-		}
-	} else {
-		boxKeys, err = ledger.LookupKeysByPrefix(lastRound, keyPrefix, maxBoxThreshold+1)
-		if err != nil {
-			return internalError(ctx, err, errFailedLookingUpLedger, v2.Log)
-		}
-		if uint64(len(boxKeys)) > maxBoxThreshold {
-			v2.Log.Info("MaxAPIBoxPerApplication limit %d exceeded", maxBoxThreshold)
-			return ctx.JSON(http.StatusBadRequest, generated.ErrorResponse{
-				Message: "Result limit exceeded",
-				Data:    nil,
-			})
+	algodSupportsUnlimitedResults := maxBoxThreshold == 0
+
+	maxKeys := func() uint64 {
+		noRequestProvidedLimit := castedMax == 0
+		dominatedByQryParams := castedMax > 0 && (maxBoxThreshold >= castedMax || algodSupportsUnlimitedResults)
+		returnsAll := noRequestProvidedLimit && algodSupportsUnlimitedResults
+
+		if dominatedByQryParams || returnsAll {
+			return castedMax
 		}
+		return maxBoxThreshold + 1
+	}
```

Comment:
```suggestion
	maxKeys := func() uint64 {
	  if castedMax == 0 {
	    if maxBoxThreshold == 0 {
	      return 0
	    }
	    return maxBoxThreshold + 1
	  }
	  if castedMax <= maxBoxThreshold {
	    return castedMax
	  }
	  return maxBoxThreshold + 1
	}
```

Reply:
... feels like this deserves a test of some kind

Reply:
@tzaffi A couple of responses:
* `TestBoxNamesByAppID` from https://github.com/algorand/go-algorand/blob/bab6d99487b3e2443b2e6b7d8237617164ca1d85/test/e2e-go/restAPI/restClient_test.go#L1234 exercises most of the functionality.
* Let me know - If after reviewing the above test you'd like to see additional testing, I can make `maxKeys` into a private function and write unit tests.  
* For the provided _suggestion_, I don't have a clear evaluation rubric.  If you feel strongly that it's an improvement, I'll take it in.
  * I think the _suggestion_ presents a different style.  As far as I can tell, it's _not_ about correctness.
  * The PR already makes one stylistic refactoring attempt.  I thought what I wrote is clearer.  Ultimately, I don't have a strong opinion and prefer to reach a conclusion.

Reply:
* Regarding additional testing. The cases `castedMax == 0` and `castedMax == 10` are tested, but it is difficult to mock `maxBoxThreshold` apparently as I don't see occurrences in that file. So yes, I recommend creating a private function and testing as this logic is important to validate.
* As for the provided suggestion, as you note, the differences are stylistic, so there is no obvious rubric. One could argue that with no new variables and fewer Boolean clauses (3 vs. 6) inside of `maxKeys()` there is less that can go wrong in future refactorings and the truth table logic is easier to understand, but the counterargument for the original code is it that it is self-commenting.
* If you agree to the suggestion, I'm happy to take over the child PR and break out a private function with a unit test

Reply:
Following up based on a verbal with @tzaffi:

* https://github.com/algorand/go-algorand/pull/4543/commits/55cca36bf0b12691c8fa2869185cc7694b283fc9 adds unit tests with the as was approach.
* Per request, I put aside my preferences and took the _suggestion_.  The _suggestion_ is slightly modified to account for new variable names.  I also found + fixed a bug.


@tzaffi Can you double check if you agree with the changes I've implemented?

Reply:
These don't look like the changes that I was suggesting which I'm re-printing below. But it's not a big deal and I'm cognizant of our release schedule so will approve any functionality backed by a fully covering unit tests.

```go
	maxKeys := func() uint64 {
	  if castedMax == 0 {
	    if maxBoxThreshold == 0 {
	        return 0 // unlimited results when both castedMax and maxBoxThreshold are 0
	      }
	      return maxBoxThreshold + 1 // API limit dominates
	  }
	  if castedMax <= maxBoxThreshold { 
	    return castedMax // user provided limit dominates
	  }
	  return maxBoxThreshold + 1 // API limit dominates
	}
```

Reply:
NVM - I somehow was looking at stale code.